src/packet/bwe/mod.rs

```diff
@@ -231,3 +240,149 @@ impl fmt::Display for BandwidthUsage {
         }
     }
 }
+
+/// Sliding window [`SeqNo`]s storage.
+///
+/// Only remembers the last [`SIZE`] packets added.
+#[derive(Debug)]
+struct HandledPacketsTracker<const SIZE: usize> {
+    /// Recently added packets.
+    history: [Option<u16>; SIZE],
+
+    /// Queue that tracks added packets order so older packets are removed.
+    queue: VecDeque<u16>,
+}
+
+impl<const SIZE: usize> HandledPacketsTracker<SIZE> {
+    /// Remembers the give [`SeqNo`].
+    ///
+    /// Expects somewhat sequential data with reordering no more than configured
+    /// [`SIZE`].
+    fn add(&mut self, seq: SeqNo) {
+        let seq = seq.as_u16();
+        let history_idx = seq as usize % SIZE;
+
+        self.queue.push_back(seq);
+        if self.queue.len() == SIZE {
+            let to_remove = self.queue.pop_front().unwrap();
+            let remove_idx = to_remove as usize % SIZE;
+            if self.history[remove_idx] == Some(seq) {
+                self.history[remove_idx] = None;
+            }
+        }
+        self.history[history_idx] = Some(seq);
+    }
+
+    /// Checks if provided [`SeqNo`] has been seen in the window.
+    fn contains(&self, seq: SeqNo) -> bool {
+        let seq = seq.as_u16();
+        let history_idx = seq as usize % SIZE;
+
+        self.history[history_idx] == Some(seq)
+    }
+}
+
+impl<const SIZE: usize> Default for HandledPacketsTracker<SIZE> {
+    fn default() -> Self {
+        Self {
+            history: [None; SIZE],
+            queue: VecDeque::with_capacity(SIZE),
+        }
+    }
+}
```

Comment:
Is this code from libWebRTC? It looks a bit dodgy wrt to rollover. Seems the internals could work off SeqNo instead of going to `u16`.

Reply:
As we are clipping to a smaller space(256) I don't think it matters. For example, `65535 % 256 = 255`, `0 % 256 = 0` and `65535 % 256 = 255`, `65536 % 256 = 0`.